src/scheduler/buckets.c

```diff
@@ -0,0 +1,1249 @@
+/*
+ * Copyright (C) 1994-2018 Altair Engineering, Inc.
+ * For more information, contact Altair at www.altair.com.
+ *
+ * This file is part of the PBS Professional ("PBS Pro") software.
+ *
+ * Open Source License Information:
+ *
+ * PBS Pro is free software. You can redistribute it and/or modify it under the
+ * terms of the GNU Affero General Public License as published by the Free
+ * Software Foundation, either version 3 of the License, or (at your option) any
+ * later version.
+ *
+ * PBS Pro is distributed in the hope that it will be useful, but WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Commercial License Information:
+ *
+ * For a copy of the commercial license terms and conditions,
+ * go to: (http://www.pbspro.com/UserArea/agreement.html)
+ * or contact the Altair Legal Department.
+ *
+ * Altair’s dual-license business model allows companies, individuals, and
+ * organizations to create proprietary derivative works of PBS Pro and
+ * distribute them - whether embedded or bundled with other software -
+ * under a commercial license agreement.
+ *
+ * Use of Altair’s trademarks, including but not limited to "PBS™",
+ * "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+ * trademark licensing policies.
+ *
+ */
+
+#include <pbs_config.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#include <log.h>
+#include "data_types.h"
+#include "pbs_bitmap.h"
+#include "node_info.h"
+#include "server_info.h"
+#include "buckets.h"
+#include "globals.h"
+#include "resource.h"
+#include "resource_resv.h"
+#include "simulate.h"
+#include "misc.h"
+#include "sort.h"
+#include "node_partition.h"
+#include "check.h"
+
+/* bucket_bitpool constructor */
+bucket_bitpool *
```

Comment:
I usually don't for constructors, destructors, and copy constructors.  The function header will be quite boring since all constructor headers will be pretty much the same.  If you still want me to add them, I can but I'd rather not.